C3-713: Redesign public representation for certified machines

andrejvelichkovski opened the pull request:
## Done

- Updated the design for each certified machine on the public website:
    - Motivation: the older design was outdated and not very representative for our work
- Implemented new view for presenting platforms. Platform is a collection of multiple machines, sharing the base components. Each platform might have more than one machine belonging to it (see example screenshots below).

The implementation is based on this design:

![Screenshot from 2024-08-20 16-57-03](https://github.com/user-attachments/assets/dfc23190-d834-49ef-b3a0-6eb7c5073c57)

## Demo examples

* Machine representation: https://ubuntu-com-14135.demos.haus/certified/202212-31031
* Platform representation: https://ubuntu-com-14135.demos.haus/certified/platforms/14169

## QA

- Check out this feature branch
- Run the site using the command `./run serve` or `dotrun`
- View the site locally in your web browser at: http://0.0.0.0:8001/
    - Be sure to test on mobile, tablet and desktop screen sizes
- [List additional steps to QA the new features or prove the bug has been resolved]

## Issue / Card

Fixes [C3-713](https://warthogs.atlassian.net/browse/C3-713)

## Screenshots

### Machine page
![Screenshot from 2024-08-20 14-46-06](https://github.com/user-attachments/assets/3a1eafca-6076-441f-93b5-d5c539108802)

### Platform page
![Screenshot from 2024-08-20 14-46-20](https://github.com/user-attachments/assets/004d33fe-d112-4c36-9976-b8b065e92085)

## Help

[QA steps](https://discourse.canonical.com/t/qa-steps/152) - [Commit guidelines](https://discourse.canonical.com/t/commit-guidelines/148)


[C3-713]: https://warthogs.atlassian.net/browse/C3-713?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

webteam-app commented:
andrejvelichkovski is not a collaborator of the repo

nadzyah commented:
The `test_model_details` test fails because we're changing the API as well since we need more information from the server to implement the design.

akbarkz commented:
The demo instance link: https://ubuntu-com-14135.demos.haus/

andrejvelichkovski commented:
Converting this PR to draft as we are still awaiting on feedback from different partners.

juanruitina commented:
Glad to see it in such good shape. Just very minor comments, all for the machine pages:

- [ ] Change the button text for the vendor link to "Visit the vendor website" to make it clear what we are doing here
- [ ] In [here](https://ubuntu-com-14135.demos.haus/certified/202212-30999), can you correctly capitalise "Notebook" in the hero H2?
- [ ] Please only show the tabs if there's more than one release. (It's OK to still show them in the platform page, it's informative there; here it's redundant with the H2)
- [ ] Change the second column in the two spec tables from `colspan="7"` to `colspan="6"`, that way they will align nicely with the grid and the "Issues? Let us know" section.
- [ ] To make the page lighter, I think we can do without the "Notes" heading, see below. In the end, the paragraph above is also a note. (Please also remove `u-no-margin--bottom` from the paragraph above)

![Screenshot 2024-09-04 at 13 43 23](https://github.com/user-attachments/assets/f41fb58a-a5bc-43a1-b8d6-cdafd0b68eb5)

+1ing this already as to not block.

andrejvelichkovski commented:
Thanks for the review @juanruitina

> * [x]  Change the button text for the vendor link to "Visit the vendor website" to make it clear what we are doing here
Sorry, I missed this, changed it now

> * [x]  In [here](https://ubuntu-com-14135.demos.haus/certified/202212-30999), can you correctly capitalise "Notebook" in the hero H2?
This is a bit more complicated issue. To somewhat resolve it, I added a `capitalize` filter that should do this. However, the root cause is that a wrong data gets exposed from our APIs. We have an open ticket to fix this, so hopefully when we get that done the data itself will be in much better shape.

> * [x] Please only show the tabs if there's more than one release. (It's OK to still show them in the platform page, it's informative there; here it's redundant with the H2)
Great suggestion, I now added an if clause to check this

> * [x]  Change the second column in the two spec tables from `colspan="7"` to `colspan="6"`, that way they will align nicely with the grid and the "Issues? Let us know" section.
Done.

> * [x]  To make the page lighter, I think we can do without the "Notes" heading, see below. In the end, the paragraph above is also a note. (Please also remove `u-no-margin--bottom` from the paragraph above)
Done.

andrejvelichkovski commented:
> @andrejvelichkovski Could you please rebase your branch with main, this will help resolve the dependency issues and hence causing error in demo

Thanks @immortalcodes, just rebased with main.

lyubomir-popov commented:
Mostly looks good, a few nit picks:

- generally, it's best to apply spacing at the bottom of elements, not at the top. In this case, you should use a p-section--shallow on the preceding element, rather than adding 50px margin-top. (also, 50px doesn't align to the baseline grid, it's always best to use our spacing variables instead)

![image](https://github.com/user-attachments/assets/2d3aa5b3-f5c8-49ac-a4e8-3ba11287d5f1)

these hrs that don't span the full width should have the class "is-muted" on them. I've asked the vanilla team to update the guidance on when to use muted horizontal rules, as I feel it is not quite clear enough currently. You ca nfind my proposed wording on [this issue](https://github.com/canonical/vanilla-framework/issues/5357), soon to be added to the documentation.

- I suggest using the [25/75 layout](https://vanillaframework.io/docs/examples/patterns/grid/25-75?theme=light) on medium as well, so we can keep the laptop icon and ubuntu certified badge visible on medium:
before:
![image](https://github.com/user-attachments/assets/55896833-c8a8-4b95-8506-8f8674974d2d)
after:
![image](https://github.com/user-attachments/assets/573a5663-3cc6-4c75-858a-6dc0361fec61)
would also apply the sdame thinking here:
![image](https://github.com/user-attachments/assets/8482862c-c649-4e1e-9e61-d1c8fd789f02)

- please remove all instances of u-sv3 and other spacing utilities and use p-section--shallow where needed. For these links specifically, I don't think it is needed. Just a muted hr in between would suffice.

- when it comes to the pagination, is this what we expect @juanruitina?
![image](https://github.com/user-attachments/assets/a6987b9a-485d-433f-87b7-8c2cb1bea135)

andrejvelichkovski commented:
Thanks a lot for the review @lyubomir-popov. I think I now addressed your concerns, but wasn't fully sure about few things.

1. As suggested I don't hide the icon on medium sizes now. However, I had one issue that I had to define empty div column to make the top breadcrumbs fit into a 25-75 vanilla grid. I noticed that by default Vanilla orders 6 columns in medium, which can't be split into 25-75 ratio, and because of this I couldn't use the empty columns like `col-start-{size}-{index}`.
2. I modified all short hrs to be muted.
3. I removed all `u-sv` instances
4. I removed the manually specified margin and used p-section--shallow as suggested.

Still, I'm a bit unsure about the margins and how everything fits together. Can you please check the updated contents on these different configurations, I think they cover all the cases the might get presented on the website:
* Classic example: https://ubuntu-com-14135.demos.haus/certified/202111-29635
* Classic platform example: https://ubuntu-com-14135.demos.haus/certified/platforms/12764
* Machine with website: https://ubuntu-com-14135.demos.haus/certified/202212-30999
* Machine with components: https://ubuntu-com-14135.demos.haus/certified/201903-26908
* Machine with notes: https://ubuntu-com-14135.demos.haus/certified/202207-30417
* More unique SoC machine: https://ubuntu-com-14135.demos.haus/certified/201710-25838 (also an example where we don't show the text 1 more certified configuration... below the Hardware details link)

In terms of the pagination, it was in my understanding that it is supposed to work that way, but I'm happy to change it.

anthonydillon commented:
@immortalcodes could you complete the code and QA review here please as you previously commented.

immortalcodes reviewed (changes requested):
I have left couple of comments, most are valid for whole of the code.
Please feel free to reach out for more clarification.
I have mostly pointed out major parts that may need a bit of refactoring.

lyubomir-popov commented:
- can we please take off the margin of the tablist:

![image](https://github.com/user-attachments/assets/e757f5ab-29a1-4c9e-98b2-40117af7fb01)

- do we need to say 2-2, 3-3? why not just 2 of 4, 3 of 4
 
![image](https://github.com/user-attachments/assets/0ddd2fc6-dc6c-4976-ba1e-7103ffed7987)

- can we align the table with the content above?
![image](https://github.com/user-attachments/assets/b7ab01ac-bb56-478a-a6e9-9e8c5229e984)

- why u-no-padding--top? 
![image](https://github.com/user-attachments/assets/1ce05582-6955-43f7-84ab-8cc648048850)

andrejvelichkovski commented on templates/certified/model-details.html:
```diff
@@ -8,322 +8,344 @@
+<section class="u-no-margin--bottom row">
+  <section class="p-section--hero">
+    <div class="row--25-75">
+      <div class="col"></div>
+      <div class="col">
+        <h1 class="p-muted-heading">
```
I actually had one issue with this. Namely, I initially used the grid. But vanilla grid by default puts 12 elements on large screens. 12 columns are easily separable into 25-75 ratio. I initially used this approach.

However, @lyubomir-popov recommended we keep the 25-75 ratio split on medium screens as well. I initially wanted to use the same grid, but I noticed that vanilla puts 6 grids on a medium sized screen.

So, I instead decided to use the `row--25-75` decorator which applies to medium screen as well. This meant having an empty `col` as well.

I'm not super experienced with Vanilla, so I couldn't figure out a better approach.

andrejvelichkovski replied:
Oh, my bad, I didn't notice that row-27-75 is actually 33-66 on medium screens. I will fix this.

andrejvelichkovski commented on templates/certified/model-details.html:
```diff
@@ -8,322 +8,344 @@
+{% if components %}
+<section class="row p-section--deep">
+  <hr/>
+  <div class="u-fixed-width">
```
Good catch, this is an old code that we just modified to use the new style guide. I now moved it to a separate file as it is a completely separate logic from the rest.

andrejvelichkovski commented on templates/certified/model-details.html:
```diff
@@ -8,322 +8,344 @@
+    <div class="row--25-75-on-medium row--25-75-on-large">
+      <div class="col u-hide--small">
+        {{ image(url="https://assets.ubuntu.com/v1/d9c2c5bf-ubuntu_certified.png", alt="Logo of the Ubuntu Certified project", width="112", height="157", hi_def=True, loading="auto") | safe }}
+      </div>
```
Thanks for sharing this doc! According to it, I believe the best description for it would be "Ubuntu Hardware Certification" to which I updated the code.

andrejvelichkovski commented on templates/certified/model-details.html:
```diff
@@ -8,322 +8,344 @@
+
+{# djlint: off #}
+<section class="u-no-margin--bottom row">
+  <section class="p-section--hero">
```
I have now removed the duplicated section.

andrejvelichkovski commented on templates/certified/model-details.html:
```diff
@@ -8,322 +8,344 @@
+.p-tabs__list {
+  margin: 0 auto;
+}
+
```
I moved this to a separate file: `_pattern_certified-model-details-tabs.scss`. From my testing it seems like it works fine. Can you check this as well?

andrejvelichkovski commented on templates/certified/hardware-details.html:
```diff
@@ -15,50 +15,9 @@
-        ) | safe
-      }}
-      {% endif %}
-      {% if release_details.category == "Server" %}
-      {{ image (
-        url="https://assets.ubuntu.com/v1/fdf83d49-Server.svg",
-        alt="Server",
-        width="80",
-        height="96",
-        hi_def=True,
-        loading="auto"
-        ) | safe
-      }}
-      {% endif %}
-      {% if release_details.category == "Ubuntu Core" %}
-      {{ image (
-        url="https://assets.ubuntu.com/v1/f9f8ace9-gateway.svg",
-        alt="Gateway",
-        width="96",
-        height="100",
```
As advised above, I now re-formatted this using djlint.

andrejvelichkovski commented:
Thanks for the review @lyubomir-popov. I think I have now addressed the points:
> why u-no-padding--top?

I was under the impression that according to the design we want the top line of H1 to match with the top line of the category image. But this doesn't seem to be the case. I now removed this class.

> can we align the table with the content above?

As we are using a custom pagination here, I have updated it to match with the Vanilla default values.

> do we need to say 2-2, 3-3? why not just 2 of 4, 3 of 4

I have now updated the logic for presenting this text that if the `start === end` we only show one value.

> can we please take off the margin of the tablist:

I have now remove the bottom margin from the list.

immortalcodes commented:
LGTM! Thanks for showing great hardwork and resilience on this PR.
I also has updated the demo.
:tada:

lyubomir-popov commented:
lgtm!
